flycheck.el

```diff
@@ -2792,6 +2837,62 @@ because some users override that function, as described in URL
   (flycheck-clear-idle-change-timer)
   (flycheck-buffer-automatically 'idle-change))
 
+(defvar-local flycheck-idle-buffer-switch-timer nil
+  "Timer to mark the idle time since switching to a new buffer.")
+
+(defvar-local flycheck--force-checking-intermediate-buffers nil
+  "If non-nil, don't cancel syntax checks for intermediate buffers.
+Ignore `flycheck-buffer-switch-check-intermediate-buffers' in
+that case.")
+
+(defun flycheck-clear-idle-buffer-switch-timer ()
+  "Clear the idle buffer switch timer."
+  (when flycheck-idle-buffer-switch-timer
+    (cancel-timer flycheck-idle-buffer-switch-timer)
+    (setq flycheck-idle-buffer-switch-timer nil)))
+
+(defvar flycheck--last-buffer (current-buffer)
+  "The current buffer or the buffer that was previously current.
+This is usually equal to the current buffer, unless the user just
+switched buffers.  After a buffer switch, it is the previous
+buffer.")
+
+(defun flycheck-handle-buffer-switch ()
+  "Handle a possible switch to another buffer.
+
+If a buffer switch actually happened, schedule a syntax check."
+  (unless (equal flycheck--last-buffer (current-buffer))
+    (setq flycheck--last-buffer (current-buffer))
+    (when flycheck-mode
+      ;; Reschedule any check that was going to be run in the current
+      ;; buffer due to a buffer switch.
+      (flycheck-clear-idle-buffer-switch-timer)
+      ;; If we can guarantee that this timer is actually going to
+      ;; result in a syntax check, we can cancel any other
+      ;; pre-existing timers that would cause syntax checks.
+      (when (memq 'idle-buffer-switch flycheck-check-syntax-automatically)
+        ;; Make sure that this timer can't get canceled by the user
+        ;; switching to another buffer.
+        (setq flycheck--force-checking-intermediate-buffers t)
+        (flycheck-clear-idle-change-timer))
+      (setq flycheck-idle-buffer-switch-timer
+            (run-at-time flycheck-idle-buffer-switch-delay nil
+                         #'flycheck-handle-idle-buffer-switch
+                         (current-buffer))))))
+
+(defun flycheck-handle-idle-buffer-switch (buffer)
+  "Handle an expired idle timer in BUFFER since the last buffer switch."
+  (when (buffer-live-p buffer)
+    (with-current-buffer buffer
+      (flycheck-clear-idle-buffer-switch-timer)
+      (when (or flycheck-buffer-switch-check-intermediate-buffers
+                flycheck--force-checking-intermediate-buffers
+                (equal buffer (current-buffer)))
+        ;; Setting this variable to non-nil guarantees exactly one
+        ;; syntax check.  After that, we can remove the effects.
+        (setq flycheck--force-checking-intermediate-buffers nil)
```

Comment:
Could you elaborate on the role of this variable?  I'm not sure to follow what it does.

Reply:
Basically, the point of that variable is to ensure that the interaction between the two timers never results in a situation where the user edits a buffer, but a syntax check does not result.

That could happen in the following situation:

1. I edit buffer A. That starts `flycheck-idle-change-timer` in buffer A.
2. I switch to buffer B. That starts `flycheck-idle-buffer-switch-timer` in buffer B.
3. I switch back to buffer A. That starts `flycheck-idle-buffer-switch-timer` also in buffer A. We only want to trigger one syntax check, so that also cancels `flycheck-idle-change-timer` in buffer A.
4. Now suppose I switch back to buffer B again. That resets `flycheck-idle-buffer-switch-timer` in buffer B.
5. When `flycheck-idle-buffer-switch-timer` expires in buffer B, a syntax check is performed.
6. When `flycheck-idle-buffer-switch-timer` expires in buffer A, however, a syntax check is only performed if `flycheck-buffer-switch-check-intermediate-buffers` is non-nil. Otherwise, nothing happens. Thus, we have forgotten about the fact that the user edited buffer A.

To avoid this, we set the variable `flycheck--force-checking-intermediate-buffers` after canceling `flycheck-idle-change-timer` in step 3. Then in step 6, the fact that the variable is non-nil forces a syntax check to be performed, regardless of the value of `flycheck-buffer-switch-check-intermediate-buffers`.